api/admin/controller/__init__.py

```diff
@@ -372,6 +374,63 @@ def __call__(self, collection, book, path=None):
         response.set_cookie("csrf_token", csrf_token, httponly=True)
         return response
 
+class TimestampsController(AdminCirculationManagerController):
+    """Returns a dict: each key is a type of service (script, monitor, or coverage provider);
+    each value is a nested dict in which timestamps are organized by service name and then by collection ID."""
+
+    def diagnostics(self):
+        self.require_system_admin()
+        timestamps = self._db.query(Timestamp).order_by(Timestamp.start)
+        if isinstance(timestamps, ProblemDetail):
+            return timestamps
+        sorted = self._sort_by_type(timestamps)
+        for type in sorted.keys():
+            for service in sorted[type]:
+                by_collection = self._sort_by_collection(sorted[type][service])
+                sorted[type][service] = by_collection
+
+        return sorted
+
+    def _sort_by_type(self, timestamps):
+        """Takes a list of Timestamp objects.  Returns a dict: each key is a type of service
+        (script, monitor, or coverage provider); each value is a dict in which the keys are the names
+        of services and the values are lists of timestamps."""
+
+        result = {}
+        for type in ["script", "monitor", "coverage_provider"]:
```

Comment:
You can use the `Timestamp` constants here. We don't have a comprehensive list inside `Timestamp`, which would be better, but if we ever add a new type of `Timestamp` we can create one then.

Reply:
On second look, it shouldn't be necessary to explicitly list the types in here at all. You can iterate over `timestamps` and put each one into a dictionary of lists, keyed off of `ts.service_type`. Something like this should work:

```
result = {}
for ts in timestamps:
    info = self._extract_info(ts)
    result.setdefault(ts.service_type, []).append(info)
```

You don't need to make a comprehensive list of all the types of Timestamps, and you only need to iterate over `timestamps` once.